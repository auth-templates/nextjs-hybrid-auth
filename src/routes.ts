export const PrivateRoutes = {
    dashboard: '/dashboard',
    account: '/account',
    leaves: '/leaves'
};

export const PublicRoutes = {
    login: '/login',
    register: '/register',
    recoverPassword: '/recover-password',
    resendConfirmationEmail: '/resend-confirmation-email',
    resetPassword: '/reset-password/:token',
    verifyAccount: '/verify-account/:token',
    twoStep: '/two-step',
}