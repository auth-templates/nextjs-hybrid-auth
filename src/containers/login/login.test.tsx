import { render, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import renderer from 'react-test-renderer';
import LoginContainer from './login';
import { PrivateRoutes } from '../../routes';
import { MemoryRouter } from 'react-router-dom';

const mockedUsedNavigate = jest.fn();
jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useNavigate: () => mockedUsedNavigate,
}));

describe('LoginContainer', () => {
    test('it should navigate to dashboard view if login request is succesful', async () => {
        render(        
            <MemoryRouter>
                <LoginContainer />
            </MemoryRouter>
        );
        const email = screen.getByLabelText('Email');
        userEvent.type(email, "pass@mail.com");
        const password = screen.getByLabelText('Password');
        userEvent.type(password, "password");
        const loginButton = screen.getByRole('button', { name: 'Login'});
        userEvent.click(loginButton);
        await waitFor(() => {
            expect(mockedUsedNavigate).toBeCalledTimes(1);
            expect(mockedUsedNavigate).toBeCalledWith(PrivateRoutes.dashboard);
        })
    });

    test('it should display an error message when csrf token is a not valid', async () => {
        render(        
            <MemoryRouter>
                <LoginContainer />
            </MemoryRouter>
        );
        const email = screen.getByLabelText('Email');
        userEvent.type(email, "badcsrftoken@mail.com");
        const password = screen.getByLabelText('Password');
        userEvent.type(password, "password");
        const loginButton = screen.getByRole('button', { name: 'Login'});
        userEvent.click(loginButton);

        expect(await screen.findByText('Login failed! Un unexpected error occured.')).toBeInTheDocument();
    });

    test('it should display an error message when there is a network failure', async () => {
        render(        
            <MemoryRouter>
                <LoginContainer />
            </MemoryRouter>
        );
        const email = screen.getByLabelText('Email');
        userEvent.type(email, "networkfail@mail.com");
        const password = screen.getByLabelText('Password');
        userEvent.type(password, "password");
        const loginButton = screen.getByRole('button', { name: 'Login'});
        userEvent.click(loginButton);

        expect(await screen.findByText('Login failed! Un unexpected error occured.')).toBeInTheDocument();
    });

    test('it should display an error message when email is invalid', async () => {
        render(        
            <MemoryRouter>
                <LoginContainer />
            </MemoryRouter>
        );
        const email = screen.getByLabelText('Email');
        userEvent.type(email, "invalidaccount@mail.com");
        const password = screen.getByLabelText('Password');
        userEvent.type(password, "password");
        const loginButton = screen.getByRole('button', { name: 'Login'});
        userEvent.click(loginButton);

        expect(await screen.findByText(/No account with the email/)).toBeInTheDocument();
    });

    test('it should display an error message when password is invalid', async () => {
        render(        
            <MemoryRouter>
                <LoginContainer />
            </MemoryRouter>
        );
        const email = screen.getByLabelText('Email');
        userEvent.type(email, "invalidpassword@mail.com");
        const password = screen.getByLabelText('Password');
        userEvent.type(password, "password");
        const loginButton = screen.getByRole('button', { name: 'Login'});
        userEvent.click(loginButton);

        expect(await screen.findByText(/Password is invalid/)).toBeInTheDocument();
    });

    it('renders correctly', () => {
        const tree = renderer.create(
            <MemoryRouter> 
                <LoginContainer />
            </MemoryRouter> 
        ).toJSON();
        expect(tree).toMatchSnapshot();
    });
})