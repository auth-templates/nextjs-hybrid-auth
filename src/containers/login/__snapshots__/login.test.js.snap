// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginContainer renders correctly 1`] = `
<form
  className="login"
  onSubmit={[Function]}
>
  <div
    className="group"
  >
    <label
      htmlFor="email"
    >
      Email
    </label>
    <input
      className="normal"
      id="email"
      name="email"
      onFocus={[Function]}
      required={true}
      type="email"
    />
  </div>
  <div
    className="group"
  >
    <label
      htmlFor="password"
    >
      Password
    </label>
    <div
      className="passwordInput"
    >
      <input
        className="normal input"
        id="password"
        name="password"
        onFocus={[Function]}
        required={true}
        type="password"
      />
      <button
        className="visibilityIcon"
        onClick={[Function]}
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="VisibilityOutlinedIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M12 6c3.79 0 7.17 2.13 8.82 5.5C19.17 14.87 15.79 17 12 17s-7.17-2.13-8.82-5.5C4.83 8.13 8.21 6 12 6m0-2C7 4 2.73 7.11 1 11.5 2.73 15.89 7 19 12 19s9.27-3.11 11-7.5C21.27 7.11 17 4 12 4zm0 5c1.38 0 2.5 1.12 2.5 2.5S13.38 14 12 14s-2.5-1.12-2.5-2.5S10.62 9 12 9m0-2c-2.48 0-4.5 2.02-4.5 4.5S9.52 16 12 16s4.5-2.02 4.5-4.5S14.48 7 12 7z"
          />
        </svg>
      </button>
    </div>
  </div>
  <div
    className="actionsGroup"
  >
    <div
      className="subgroup"
    >
      <button
        className="normal loginButton"
        type="submit"
      >
        Login
      </button>
      <span
        className="forgotPasswordQuestion"
      >
        <a
          className="link"
          href="/recover-password"
          onClick={[Function]}
        >
           Forgot your password? 
        </a>
      </span>
    </div>
    <div
      className="subgroup"
    >
      <span>
         Don't have an account? 
        <a
          className="link"
          href="/sign-up"
          onClick={[Function]}
        >
           Sign Up 
        </a>
      </span>
      <span>
        <a
          className="link"
          href="/resend-confirmation-email"
          onClick={[Function]}
        >
           Resend confirmation email 
        </a>
      </span>
    </div>
  </div>
</form>
`;
